File: 0145_Binary_Tree_Postorder_Traversal/MQQM2.cpp
/*
  题目：
  给定一个二叉树，返回它的 后序 遍历。
  
  示例:
  输入: [1,null,2,3]  
     1
      \
       2
      /
     3 
  输出: [3,2,1]
  
  进阶: 递归算法很简单，你可以通过迭代算法完成吗？
  
  非递归做法：
  后序遍历是：左-右-根，颠倒下顺序（可以用栈实现）就是：根-右-左。这就可以参考“先序遍历的非递归实现”。
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {        
        stack<TreeNode*> res; // res这个栈 全pop出来 就是后序遍历
        
        stack<TreeNode*> stk;
        TreeNode* p = root;        
        while( !stk.empty() || p!=NULL ){            
            while( p!=NULL ){                 
                res.push(p);
                stk.push(p);              
                p = p->right;                
            }
            p = stk.top()->left;
            stk.pop();                
        }        
        
        vector<int> ret;
        while( !res.empty() ){
            ret.push_back( res.top()->val );
            res.pop();            
        }       
        return ret;
    }
};

File: 0020_Valid_Parentheses/MQQM2.go
/*
  题目：
  给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。
  
  说明：
  有效字符串需满足：左括号必须用相同类型的右括号闭合。左括号必须以正确的顺序闭合。
  注意空字符串可被认为是有效字符串。
*/
func isValid(s string) bool {
	if len(s) == 0 {
		return true
	}

	m := map[byte]byte{
		'(': ')',
		'[': ']',
		'{': '}',
	}

	var stk []byte
	for i := 0; i < len(s); i++ {
		si := s[i]
		if si == '(' || si == '[' || si == '{' {
			stk = append(stk, si)
			continue
		}
		if si == ')' || si == ']' || si == '}' {
			if len(stk) == 0 || m[stk[len(stk)-1]] != si {
				return false
			} else {
				stk = stk[:len(stk)-1]
			}
		}
	}
	
	if len(stk) == 0 {
		return true
	}

	return false

}
